src/ray/raylet/node_manager.cc

```diff
@@ -735,6 +739,83 @@ void NodeManager::DispatchTasks(
+    const WorkerID &worker_id) {
+  bool is_worker = true;
+  auto worker = worker_pool_.GetRegisteredWorker(worker_id);
+  if (!worker) {
+    is_worker = false;
+    worker = worker_pool_.GetRegisteredDriver(worker_id);
+    if (!worker) {
+      return std::make_pair(nullptr, true);
+    }
+  }
+  return std::make_pair(worker, is_worker);
+}
+
+#define PREPROCESS_REQUEST(REQUEST_TYPE, REQUEST, SEND_REPLY)                       \
+  WorkerID worker_id = WorkerID::FromBinary(REQUEST.worker_id());                   \
+  do {                                                                              \
+    if (!PreprocessRequest<rpc::REQUEST_TYPE>(worker_id, #REQUEST_TYPE)) {          \
+      SEND_REPLY(                                                                   \
+          Status::Invalid("Discard this request due to failure of preprocessing."), \
+          nullptr, nullptr);                                                        \
+      return;                                                                       \
+    }                                                                               \
+  } while (0)
+
+template <typename Request>
+bool NodeManager::PreprocessRequest(const WorkerID &worker_id,
+                                    const std::string &request_name) {
+  std::ostringstream oss;
+  if (RAY_LOG_ENABLED(DEBUG)) {
+    oss << "Received a " << request_name << " request. Worker id " << worker_id << ".";
+  }
+  // NOTE(Joey Jiang): We check heartbeat actively because the heartbeats' callback won't
+  // be invoked until all the posted requests in the io_service have been finished.
+  // The raylet might be marked as dead due to missing heartbeats when too many requests
+  // are sent to this raylet.
+  int64_t expiry_delay =
+      std::chrono::duration_cast<std::chrono::milliseconds>(
+          std::chrono::steady_clock::now() - heartbeat_timer_.expires_at())
+          .count();
+  int64_t expiry_delay_limit = RayConfig::instance().heartbeat_timeout_milliseconds() *
+                               RayConfig::instance().num_heartbeats_timeout() / 2;
+
+  // Timer's callback hasn't been called on the desired time.
+  if (expiry_delay > expiry_delay_limit) {
+    heartbeat_timer_.cancel();
+    RAY_LOG(WARNING)
+        << "Timer has been cancelled. Maybe the timer was delayed due to "
+           "lots of post events and we have reset the heartbeat. Expiry delay "
+        << expiry_delay << ", delay limit " << expiry_delay_limit;
+    Heartbeat();
+  }
+
+  auto rt = GetWorker(worker_id);
+  auto worker = rt.first;
+  // Worker process has been killed, we should discard this request.
+  if (!worker) {
+    RAY_LOG(WARNING) << "Worker " << worker_id
+                     << " is not found in worker pool, request will be discarded.";
+    return false;
+  }
+  if (RAY_LOG_ENABLED(DEBUG)) {
+    oss << " Is worker: " << (rt.second ? "true" : "false") << ". Worker pid "
```

Comment:
It would be great if this could also print the type of the request. Also, you can just write `RAY_LOG(DEBUG) << ...`, instead of `oss << ...; RAY_LOG(DEBUG) << oss.str()`.

Reply:
I need to print only one DEBUG log for a request.